### PR TITLE
strace -c

### PR DESCRIPTION
**Pull request by rupeshkoushik07**
## Description
Fixes # [(issue)](https://github.com/Lind-Project/lind_project/issues/345#issuecomment-1917769880)

This PR fixes the Ordering issue and signing issue and implements strace -c functionality

### Type of change

- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Any dependent changes have been added to a pull request and/or merged in other modules (lind-project, lind-glibc, safeposix-rust)

### CODE REVIEW

File: src/trusted/service_runtime/nacl_syscall_strace.c [hunk 1]
```diff
@@ -6,17 +6,36 @@
 #include <stdio.h>
 #include <stdint.h>
 #include <sys/poll.h>
+#include "native_client/src/trusted/service_runtime/include/bits/nacl_syscalls.h"
```
Comment (yizhuoliang) [inline src/trusted/service_runtime/nacl_syscall_strace.c:9]: Just curious, what is this include for

File: src/trusted/service_runtime/nacl_syscall_strace.c [hunk 2]
```diff
@@ -69,444 +97,1271 @@ char* formatStringArgument(const char *input) {
...
+    }
+    
 }
 
-void NaClStraceOpen(int cageid, char* path, int flags, int mode, int fd) {
-    fprintf(tracingOutputFile, "%d open(%s, %d, %d) = %d\n", cageid, path, flags, mode, fd);
+
+
+void NaClStraceOpen(int cageid, char* path, int flags, int mode, int fd, long long elapsedTime) {
+    if (strace_C){
```
Comment (yizhuoliang) [inline src/trusted/service_runtime/nacl_syscall_strace.c]: The indentation is still messy

File: src/trusted/service_runtime/nacl_syscall_strace.c (hunk 1 above)
Comment (rennergade) [reply src/trusted/service_runtime/nacl_syscall_strace.c:9]: its for the the syscalls defines, this is good

File: src/trusted/service_runtime/nacl_syscall_strace.c [hunk 3]
```diff
@@ -6,17 +6,36 @@
 #include <stdio.h>
 #include <stdint.h>
 #include <sys/poll.h>
+#include "native_client/src/trusted/service_runtime/include/bits/nacl_syscalls.h"
 #include "native_client/src/trusted/service_runtime/nacl_syscall_strace.h"
+#include <time.h>
+#include <stdbool.h>
+#define NUM_SYSCALLS 200
```
Comment (rennergade) [inline src/trusted/service_runtime/nacl_syscall_strace.c]: i think this should be 256 to give us room

File: src/trusted/service_runtime/nacl_syscall_strace.c [hunk 4]
```diff
@@ -69,444 +97,1271 @@ char* formatStringArgument(const char *input) {
...
+    
+    fprintf(tracingOutputFile, "%d pread(%d, %p, %d, %ld) = %d\n", cageid, d, buf, count, offset, ret);
+    }
+    
 }
 
-void NaClStraceWrite(int cageid, int d, void *buf, int count, int ret) {
+void NaClStraceWrite(int cageid, int d, void *buf, int count, int ret, long long elapsedTime) {
+    if (strace_C){
+    syscallStats[NACL_sys_write].count++;
```
Comment (rennergade) [inline src/trusted/service_runtime/nacl_syscall_strace.c:231]: all of these should be indented

File: src/trusted/service_runtime/sel_main.c [hunk 5]
```diff
@@ -1013,4 +1017,4 @@ int NaClSelLdrMain(int argc, char **argv) {
 
   /* NOT REACHED */
   return ret_code;
-}
+}
```
Comment (rennergade) [inline src/trusted/service_runtime/sel_main.c]: file should end with a newline

File: src/trusted/service_runtime/nacl_syscall_strace.h [hunk 6]
```diff
@@ -1,117 +1,90 @@
...
+void NaClStraceListen(int cageid, int sockfd, int backlog, int ret, long long elapsedTime) ;
+void NaClStracePoll(int cageid, uintptr_t fds, nfds_t nfds, int timeout, int retval, long long elapsedTime) ;
+void NaClStraceFcntlGet(int cageid, int fd, int cmd, int ret, long long elapsedTime) ;
+void NaClStraceFcntlSet(int cageid, int fd, int cmd, long set_op, int ret, long long elapsedTime) ;
+void NaClStraceEpollCreate(int cageid, int size, int ret, long long elapsedTime) ;
+void NaClStraceEpollCtl(int cageid, int epfd, int op, int fd, uintptr_t event, int ret, long long elapsedTime) ;
+void NaClStraceEpollWait(int cageid, int epfd, uintptr_t events, int maxevents, int timeout, int ret, long long elapsedTime) ;
+void NaClStraceSelect(int cageid, int nfds, uintptr_t readfds, uintptr_t writefds, uintptr_t exceptfds, uintptr_t timeout, int ret, long long elapsedTime) ;
+void printFinalSyscallStats();
+const char* getSyscallName(int syscallIndex);
```
Comment (rennergade) [inline src/trusted/service_runtime/nacl_syscall_strace.h]: file should end with a newline

File: src/trusted/service_runtime/nacl_syscall_strace.c [hunk 7]
```diff
@@ -69,444 +97,1271 @@ char* formatStringArgument(const char *input) {
...
+    if (strace_C){
+    syscallStats[NACL_sys_select].count++;
+    syscallStats[NACL_sys_select].totalTime += elapsedTime;
+    if (ret < 0) {
+        syscallStats[NACL_sys_select].errorCount++;
+    }
+    } else {
+        fprintf(tracingOutputFile, "%d select(%d, 0x%08"NACL_PRIxPTR", 0x%08"NACL_PRIxPTR", 0x%08"NACL_PRIxPTR", 0x%08"NACL_PRIxPTR") = %d\n", cageid, nfds, readfds, writefds, exceptfds, timeout, ret);
+    }
+}
```
Comment (rennergade) [inline src/trusted/service_runtime/nacl_syscall_strace.c]: file should end with a newline

File: src/trusted/service_runtime/nacl_syscall_strace.c [hunk 8]
```diff
@@ -69,444 +97,1271 @@ char* formatStringArgument(const char *input) {
...
+    }
+    } else {
+        fprintf(tracingOutputFile, "%d epollwait(%d, 0x%08"NACL_PRIxPTR", %d, %d) = %d\n", cageid, epfd, events, maxevents, timeout, ret);
+    } 
 }
+
+
+void NaClStraceSelect(int cageid, int nfds, uintptr_t readfds, uintptr_t writefds, uintptr_t exceptfds, uintptr_t timeout, int ret, long long elapsedTime) {
+    if (strace_C){
+    syscallStats[NACL_sys_select].count++;
```
Comment (rennergade) [inline src/trusted/service_runtime/nacl_syscall_strace.c:1273]: indent, a ton of these to fix

File: src/trusted/service_runtime/nacl_syscall_common.h [hunk 9]
```diff
@@ -446,6 +446,8 @@ int32_t NaClSysEpollCtl(struct NaClAppThread  *natp, int epfd, int op, int fd, s
 int32_t NaClSysEpollWait(struct NaClAppThread  *natp, int epfd, struct epoll_event *events, int maxevents, int timeout);
 int32_t NaClSysSelect (struct NaClAppThread *natp, int nfds, fd_set * readfds, 
                        fd_set * writefds, fd_set * exceptfds, struct timeval *timeout);
+void printFinalSyscallStats();
+long long gettimens();
 EXTERN_C_END
 
-#endif  /* NATIVE_CLIENT_SERVICE_RUNTIME_NACL_SYSCALL_COMMON_H__ */
+#endif  /* NATIVE_CLIENT_SERVICE_RUNTIME_NACL_SYSCALL_COMMON_H__ */
```
Comment (rennergade) [inline src/trusted/service_runtime/nacl_syscall_common.h]: file should end with newline

File: src/trusted/service_runtime/nacl_syscall_common.c [hunk 10]
```diff
@@ -4727,97 +5131,111 @@ int32_t NaClSysEpollCtl(struct NaClAppThread  *natp, int epfd, int op, int fd, s
...
   #ifdef TRACING
-  NaClStraceSelect(nap->cage_id, nfds, (uintptr_t)sysreadfds, (uintptr_t)syswritefds, (uintptr_t)sysexceptfds, (uintptr_t)nacltimeout, retval);
+  long long endtime = gettimens();
+  long long totaltime = endtime - starttime;
+  NaClStraceSelect(nap->cage_id, nfds, (uintptr_t)sysreadfds, (uintptr_t)syswritefds, (uintptr_t)sysexceptfds, (uintptr_t)nacltimeout, retval, totaltime);
   #endif
 
   return retval;
-}
+}
```
Comment (rennergade) [inline src/trusted/service_runtime/nacl_syscall_common.c]: file should end with newline

File: src/trusted/service_runtime/nacl_syscall_common.c [hunk 11]
```diff
@@ -1863,7 +2086,7 @@ int32_t NaClSysMmapIntern(struct NaClApp        *nap,
 
   return (int32_t) map_result;
 }
-
+//check this
```
Comment (rennergade) [inline src/trusted/service_runtime/nacl_syscall_common.c]: comment?

File: src/trusted/service_runtime/nacl_syscall_common.c [hunk 12]
```diff
@@ -1182,200 +1333,272 @@ int32_t NaClSysChmod(struct NaClAppThread *natp,
...
 }
 
+
 int32_t NaClSysSyncFileRange(struct NaClAppThread *natp,
-                     int                   fd,
-		     off_t 		   offset,
-		     off_t 		   nbytes, 
-		     uint32_t 	           flags) {
+                             int                   fd,
+                             off_t                 offset,
```
Comment (rennergade) [inline src/trusted/service_runtime/nacl_syscall_common.c]: tons of random white space

File: src/trusted/service_runtime/nacl_syscall_common.c [hunk 13]
```diff
@@ -757,188 +823,212 @@ int32_t NaClSysRead(struct NaClAppThread  *natp,
...
-  NaClStracePread(nap->cage_id, d, (void *)sysaddr, count, offset, retval);
-  #endif
+    #ifdef TRACING
+    long long endtime = gettimens();
+    long long totaltime = endtime - starttime;
+    NaClStracePread(nap->cage_id, d, (void *)sysaddr, count, offset, retval, totaltime);
+    #endif
 
-  return retval;
+    return retval;
```
Comment (rennergade) [inline src/trusted/service_runtime/nacl_syscall_common.c]: random whitespace

File: src/trusted/service_runtime/nacl_syscall_common.c [hunk 14]
```diff
@@ -757,188 +823,212 @@ int32_t NaClSysRead(struct NaClAppThread  *natp,
...
+                     off_t offset) { 
+  struct NaClApp *nap = natp->nap;
+  int32_t retval = -NACL_ABI_EINVAL;
+  ssize_t read_result = -NACL_ABI_EINVAL;
+  uintptr_t sysaddr;
+  size_t log_bytes;
+  char const *ellipsis = "";
 
-  if (d < 0) return -NACL_ABI_EBADF;
+
```
Comment (rennergade) [inline src/trusted/service_runtime/nacl_syscall_common.c]: random whitespace

File: src/trusted/service_runtime/nacl_syscall_common.c [hunk 15]
```diff
@@ -343,152 +348,183 @@ int32_t NaClSysGetpid(struct NaClAppThread *natp) {
...
-  lind_exit(status, nap->cage_id);
+    // First, signal any other threads in the group to exit
+    if (nap->num_threads > 1) {
+        NaClExitThreadGroup(natp);
+    }
 
-  NaClLog(1, "Exit syscall handler: %d\n", status);
-  (void) NaClReportExitStatus(nap, NACL_ABI_W_EXITCODE(status, 0));
-  NaClAppThreadTeardown(natp);
+    #ifdef TRACING
```
Comment (rennergade) [inline src/trusted/service_runtime/nacl_syscall_common.c]: random whitespace

**Review by rennergade — CHANGES_REQUESTED** [review]
Looks good theoretically. A bunch of cosmetic problems.

Add newlines, fix whitespace, and indent properly!

File: src/trusted/service_runtime/nacl_syscall_common.c [hunk 16]
```diff
@@ -3395,9 +3571,13 @@ int32_t NaClSysGetTimeOfDay(struct NaClAppThread      *natp,
   if (!NaClCopyOutToUser(natp->nap, (uintptr_t)tv, &now, sizeof(now))) {
     return -NACL_ABI_EFAULT;
   }
+
```
Comment (rennergade) [inline src/trusted/service_runtime/nacl_syscall_common.c]: lots of added newlines for no reason, remove

File: src/trusted/service_runtime/sel_main.c (hunk 5 above)
Comment (rennergade) [reply src/trusted/service_runtime/sel_main.c]: still dont have a newline here, add

File: src/trusted/service_runtime/nacl_syscall_strace.c [hunk 17]
```diff
@@ -6,507 +6,1325 @@
...
+void NaClStraceGetppid(int cageid, int pid, long long elapsedTime) {
+  if (strace_C) {
+    syscallStats[NACL_sys_getppid].count++;
+    syscallStats[NACL_sys_getppid].totalTime += elapsedTime;
+  } else {
 
-void NaClStraceGetppid(int cageid, int pid) {
     fprintf(tracingOutputFile, "%d getppid() = %d\n", cageid, pid);
-}
+  }
```
Comment (rennergade) [inline src/trusted/service_runtime/nacl_syscall_strace.c]: random whitespace

File: src/trusted/service_runtime/nacl_syscall_strace.c [hunk 18]
```diff
@@ -6,507 +6,1325 @@
...
     }
+    srcPtr++;
+  }
 
-    return output;
+  return output;
 }
-
-void NaClStraceGetpid(int cageid, int pid) {
+//noerror
```
Comment (rennergade) [inline src/trusted/service_runtime/nacl_syscall_strace.c]: comment?

File: src/trusted/service_runtime/nacl_syscall_strace.c [hunk 19]
```diff
@@ -6,507 +6,1325 @@
...
+  long long totalTime; // Total time spent in the syscall (in nanoseconds)
+  long long errorCount; // Number of errors encountered in the syscall
+}
+SyscallStats;
+long long totalSyscallsMicroseconds = 0; // Total time for all syscalls (in microseconds)
+int totalSyscallsCount = 0;
+
+SyscallStats syscallStats[NUM_SYSCALLS];
+int strace_C = 0;
+static long long totalMkdirTime = 0;
```
Comment (rennergade) [inline src/trusted/service_runtime/nacl_syscall_strace.c]: unused test variables, remove

File: src/trusted/service_runtime/nacl_syscall_strace.c [hunk 20]
```diff
@@ -6,507 +6,1325 @@
...
+
+FILE * tracingOutputFile = NULL;
+long long totalSyscallsTime = 0; // Total time for all syscalls
+
+typedef struct {
+  long long count; // Number of times the syscall was called
+  long long totalTime; // Total time spent in the syscall (in nanoseconds)
+  long long errorCount; // Number of errors encountered in the syscall
+}
+SyscallStats;
```
Comment (rennergade) [inline src/trusted/service_runtime/nacl_syscall_strace.c]: this should be up with the struct and then a newline after

**Review by rennergade — CHANGES_REQUESTED** [review]
still needs more cosmetic changes

File: src/trusted/service_runtime/nacl_syscall_strace.h [hunk 21]
```diff
@@ -1,117 +1,91 @@
...
+void NaClStraceUnlink(int cageid, char* pathname, int retval, long long elapsedTime) ;
+void NaClStraceRename(int cageid, const char *oldpath, const char *newpath, int retval, long long elapsedTime) ;
+void NaClStraceMmap(int cageid, void *start, size_t length, int prot, int flags, int d, uintptr_t offset, int retval, long long time) ;
+void NaClStraceMunmap(int cageid, uintptr_t sysaddr, size_t length, int retval, long long elapsedTime) ;
+void NaClStraceShmat(int cageid, int shmid, void *shmaddr, int shmflg, int retval, long long elapsedTime) ;
+void NaClStraceShmget(int cageid, int key, size_t size, int shmflg, int retval, long long elapsedTime) ;
+void NaClStraceShmdt(int cageid, void *shmaddr, int retval, long long elapsedTime) ;
+void NaClStraceShmctl(int cageid, int shmid, int cmd, uintptr_t bufsysaddr, int retval, long long elapsedTime) ;
+void NaClStraceSocketPair(int cageid, int domain, int type, int protocol, int *lindfds, int retval, long long elapsedTime) ;
+// void NaClStraceNanosleep(int cageid, uintptr_t req, uintptr_t rem, int ret, long long elapsedTime) ;
```
Comment (rennergade) [inline src/trusted/service_runtime/nacl_syscall_strace.h]: commented?

File: src/trusted/service_runtime/sel_main.c [hunk 22]
```diff
@@ -164,6 +164,7 @@ static void PrintUsage(void) {
           " -E <name=value>|<name> set an environment variable\n"
           " -Z use fixed feature x86 CPU mode\n"
           " -t toggle runtime statistics\n"
+          " -p [s/c], s for per-call tracing, c minics strace's -c functionality"
```
Comment (rennergade) [inline src/trusted/service_runtime/sel_main.c:167]: we didnt end up doing it this way so should edit this comment @yizhuoliang

File: src/trusted/service_runtime/nacl_syscall_strace.h (hunk 21 above)
Comment (rupeshkoushik07) [reply src/trusted/service_runtime/nacl_syscall_strace.h]: We do not need them

**Pull request merged**